docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,318 @@
+  Instead of having to re-create the redirect,
+  we can just disable it and re-enable it later.
+- Allow to add a short description.
+  It's useful to document why the redirect was created.
+
+Allow matching query arguments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We can do this in two ways:
+
+- At the DB level with some restrictions.
+  If done at the DB level,
+  we would need to have a different field
+  with just the path, and other with the query arguments normalized and sorted.
+
+  For example, if we have a redirect with the value ``/foo?blue=1&yellow=2&red=3``,
+  if would be normalized in the DB as ``/foo`` and ``blue=1&red=3&yellow=2``.
+  This implies that the URL to be matched must have the exact same query arguments,
+  it can't have more or less.
+
+  I believe the implementation described here is the same being used by Netlify,
+  since they have that same restriction.
+
+      If the URL contains other parameters in addition to or instead of id, the request doesn't match that rule.
+
+      https://docs.netlify.com/routing/redirects/redirect-options/#query-parameters
+
+- At the Python level.
+  If done at the DB level,
+  we would need to have a different field
+  with just the path, and other with query arguments.
+
+  The matching of the path would be done at the DB level,
+  and the matching of the query arguments would be done at the Python level.
+  Here we can be more flexible, allowing any query arguments in the matched URL.
+
+  We had some performance problems in the past,
+  but I believe it was mainly due to the use of regex instead of using string operations.
+  And matching the path is still done at the DB level.
+  We could limit the number of redirects that can be created with query arguments,
+  or the number of redirects in general.
+
+Don't run redirects on domains from pull request previews
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We currently run redirects on domains from pull request previews,
+this is a problem when moving a whole project to a new domain.
+
+Do we have the need to run redirects on external domains?
+They are suppose to be temporary domains.
+
+Normalize paths with trailing slashes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Currently, if users want to redirect a path with a trailing slash and without it,
+they need to create two separate redirects (``/page/`` and ``/page``).
+
+We can simplify this by normalizing the path before matching it.
+
+For example:
+
+- From: ``/page/``
+  To: ``/new/page``
+
+The from path will be normalized to ``/page``,
+and the filename to match will also be normalized before matching it.
+This is similar to what Netlify does:
+https://docs.netlify.com/routing/redirects/redirect-options/#trailing-slash.
+
+Page and exact redirects without a wildcard at the end will be normalized,
+all other redirects need to be matched as is.
+
+Improving exact redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Explicitly place the ``$rest`` placeholder in the target URL,
```

Comment:
If we are okay with introducing some backwards incompatible changes to new redirects (existing redirects will be migrated to the new syntax, so those should be fine).

We may also take this opportunity to rename `$rest` to something else, or even use `*` in the from URL.

Some suggestions:

- `/en/latest/directory/*` -> `/en/latest/new-directory/$suffix`
- `/en/latest/directory/$suffix` -> `/en/latest/new-directory/$suffix` (kind of like this this one)

`$suffix` can also be named `$end`, `$rest`, `:end`, `:rest`, `:splat`.

The syntax that Netlify and CF use is:

- `/en/latest/directory/*` -> `/en/latest/new-directory/:splat`

Reply:
I'm not familiar with `:splat` as a concept. I'm a little surprised they didn't just use `*` again. If CF & Netlify both use a specific syntax, it's probably worth reusing, and not inventing our own.

Reply:
I was also confused about what `:splat` is, and even the `:var` syntax looks unfamiliar...

My guess is that one service copied the other

- https://developers.cloudflare.com/pages/platform/redirects/#splats
- https://docs.netlify.com/routing/redirects/redirect-options/#splats

But do we want to still use that unfamiliar syntax/names? I'd be fine not following those services on this one, and use something more familiar for users.

Reply:
Gitlab pages also uses :splat :/


They even mention netlify as inspiration https://docs.gitlab.com/ee/user/project/pages/redirects.html

Reply:
Yea, I think we can just use the same thing, since it seems to be used everywhere else... It seems `splat` is a name for the `*` char, and I assume the `:splat` syntax is likely JS-inspired? 🤷

Reply:
ahh, yeah, wouldn't be surprised if this was JS inspired haha

Reply:
I'd follow the other services conventions as well and follow `*` and `:splat`. There is no need to add more confusion to this topic 😄